src/libsystemd/sd-json/sd-json.c

```diff
@@ -5527,6 +5528,28 @@ _public_ int sd_json_dispatch_id128(const char *name, sd_json_variant *variant,
         return 0;
 }
 
+_public_ int sd_json_dispatch_signal(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        int *signo = userdata;
+        uint64_t k;
+
+        assert_return(variant, -EINVAL);
+
+        if (sd_json_variant_is_null(variant)) {
+                *signo = SIGNO_INVALID;
+                return 0;
+        }
+
+        if (!sd_json_variant_is_unsigned(variant))
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not an integer.", strna(name));
+
+        k = sd_json_variant_unsigned(variant);
```

Comment:
hmm, so we nowadays when dispatching json into structures also accept numbers formatted as strings when mapping them to integers. This is done to allow clients that have issues with the json 53bit integer limitations, to talk to us. Now of course, there are no signals > 2^53bit defined, but still, if we accept strings for numbers everywhere else, we should really do that here too.

my suggestion hence: rework this so that this wraps sd_json_dispatch_int32(). That helper already deals with strings as suggested, and you just have to do that SIGNAL_VALID() check afterwards.

Compare with the 50ed3b168234fe59c3b5250031f8f368241331b2 rework I recently did, which is a very good blueprint for this.